### DISCUSSION TITLE
Database data change logging

### DISCUSSION THREAD

**Discussion by ggukggukee** [prisma/prisma#12610] [Q&A]
Hello!
I found out that in Prisma there is no database change logging or I might say triggers. Therefore I created my own solution.

EditLog schema:
```prisma
model EditLog {
  id Int @id @default(autoincrement())
  action String?
  tableName String?
  columnName String?
  columnId Int?
  oldData Json?
  newData Json?
  createdAt DateTime @default(now())
}
```

Route for edit user data 
```js
router.put('/edit', async(req, res) => {
  const { userId, type, comment, date } = req.body
  try {
    // find previous data
    const prevData = await prisma.user.findFirst({
      where: {
        id: userId
      }
    })

    // updated data
    const newData = await prisma.user.update({
      where: {
        id: userId
      },
      data: {
        type: type,
        date: date,
        comment: comment
      }
    })

    // function that compare objects and get differences
    const result = await difference(prevData, newData)

    // map through object keys and store changes to DB
    Object.entries(result).map(async([key, val], i) => {
     const editlog = await prisma.editLog.create({
       data: {
         action: 'edit', 
         tableName: 'user',
         columnId: newData.id,
         columnName: key,
         oldData: prevData[key],
         newData: newData[key]
       }
     })
    })

    // return new data to front
    return res.json(user2)
  } catch (err) {
    console.log(err)
  }
});
```

Is it a good approach for storing changes?
If this will be used in all routes for POST, PUT and DELETE methods how it will affect in app's performance?

**Comment by austincrim** [comment]
Hey @ggukggukee 👋 ,

This approach should definitely work! You might want to look into creating a middleware in your HTTP framework (or even a [Prisma middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware)) to avoid duplication.

**Comment by claysam75** [comment]
Would be interested to see the workings of that 'difference' function if possible? Also what is columnId recording exactly?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
